cpp/src/parquet/encoding.cc

```diff
@@ -1888,33 +1736,23 @@ inline int DictDecoderImpl<FLBAType>::DecodeArrow(
   }
 
   PARQUET_THROW_NOT_OK(builder->Reserve(num_values));
-  arrow::internal::BitmapReader bit_reader(valid_bits, valid_bits_offset, num_values);
 
   auto dict_values = reinterpret_cast<const FLBA*>(dictionary_->data());
 
-  if (null_count != 0) {
-    for (int i = 0; i < num_values; ++i) {
-      bool is_valid = bit_reader.IsSet();
-      bit_reader.Next();
-      if (is_valid) {
-        int32_t index;
-        if (ARROW_PREDICT_FALSE(!idx_decoder_.Get(&index))) {
-          throw ParquetException("");
-        }
-        builder->UnsafeAppend(dict_values[index].ptr);
-      } else {
-        builder->UnsafeAppendNull();
-      }
-    }
-  } else {
-    for (int i = 0; i < num_values; ++i) {
+  auto decode_value = [&](bool is_valid) {
+    if (is_valid) {
       int32_t index;
       if (ARROW_PREDICT_FALSE(!idx_decoder_.Get(&index))) {
         throw ParquetException("");
       }
       builder->UnsafeAppend(dict_values[index].ptr);
+    } else {
```

Comment:
Not related, to your change, but `dict_values[index]` should check for the index bounds.

Reply:
Will do, thank you.